Add kani::spawn and an executor to the Kani library

fzaiser opened the pull request:
### Description of changes: 

This adds an executor (scheduler for async futures) to the Kani library, thus supporting `kani::spawn` as a replacement for `tokio::spawn`.

It also includes `kani::yield_now` which is similar to `tokio::yield_now`.

### Resolved issues:

Resolves #1504 

### Call-outs:

This cannot currently be merged because linking the Kani library is broken (just like the standard library).

### Testing:

* How is this change tested? **Not yet.** Could use a similar test as #1658

* Is this a refactor change? No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval commented on library/kani/src/futures.rs:
```diff
@@ -41,3 +44,216 @@ const NOOP_RAW_WAKER: RawWaker = {
+pub struct NondeterministicScheduling;
+
+impl SchedulingStrategy for NondeterministicScheduling {
+    #[cfg(kani)]
```
I don't think there is any need for this. We only distribute the crate with `kani` enabled. Additionally, `crate::any()` should fail already if this is used outside of Kani.

fzaiser replied:
OK, I'll remove the `#[cfg]`s

celinval commented on library/kani/src/futures.rs:
```diff
@@ -41,3 +44,216 @@ const NOOP_RAW_WAKER: RawWaker = {
+    }
+
+    /// Runs the scheduler with the given scheduling plan until all tasks have completed
+    #[inline] // to work around linking issue
```
Can we get rid of most of these `[inline]` since you are blocked by the linking issue no matter what?

fzaiser replied:
Good point, done

celinval commented on tests/kani/AsyncAwait/spawn.rs:
```diff
@@ -0,0 +1,49 @@
+            yield_now().await;
+            x2.fetch_add(1, Ordering::Relaxed);
+        },
+        RoundRobin::default(),
```
don't you need import this struct?

fzaiser replied:
You're right, done

celinval commented on library/kani/src/futures.rs:
```diff
@@ -41,3 +44,216 @@ const NOOP_RAW_WAKER: RawWaker = {
+type BoxFuture = Pin<Box<dyn Future<Output = ()> + Sync + 'static>>;
+
+/// Allows to parameterize how the scheduler picks the next task to poll in `spawnable_block_on`
+pub trait SchedulingStrategy {
```
Can you please improve the documentation? How would a user implement this?

fzaiser replied:
Could you be more specific as to what information is missing? Only one method has to be implemented and its documentation should include all the user needs to know?

celinval replied:
Also, I think the Scheduler is `pub(crate)`, so we shouldn't expect users to know about it in the first place. You should document this trait as to what this does, not how. Something like:

```rust
/// Trait that determines the possible sequence of tasks scheduling for a  harness.
/// 
/// <Expand what this means for the verification task at hand. Avoid giving implementation details>
///
/// <Optional Example>
``

celinval replied:
Sounds good. Shouldn't this trait be `pub(crate)` though?

celinval commented on library/kani/src/futures.rs:
```diff
@@ -41,3 +44,216 @@ const NOOP_RAW_WAKER: RawWaker = {
+
+/// Allows to parameterize how the scheduler picks the next task to poll in `spawnable_block_on`
+pub trait SchedulingStrategy {
+    /// Picks the next task to be scheduled whenever the scheduler needs to pick a task to run next, and whether it can be assumed that the picked task is still running
```
It's not clear to me what the second parameter means here. What kind of information the scheduler has to make that decision in the first place?

fzaiser replied:
I'm sorry, I'm not quite sure what you're asking here. The second parameter is the number of tasks. Are you talking about the second part of the return value? This is just a flag indicating whether the scheduler can `kani::assume` that the returned task is still running, usually because the task was chosen nondeterministically. This cuts off "useless" paths where a completed task is polled again, and is even necessary to make things terminate, if nondeterminism is involved: if we pick the task nondeterministically, and don't have the restriction to still running tasks, we could poll the same task over and over again. However, for the round robin scheduling strategy, assuming that the picked task is still running is generally not possible, and amounts to `assume(false)`, which is undesirable.

celinval replied:
Sorry, I was talking about the return value. What does the `bool` part of the return value mean?

fzaiser replied:
This is just a flag indicating whether the scheduler can `kani::assume` that the returned task is still running. As a rule of thumb: if the scheduling strategy picks the next task nondeterministically (using `kani::any()`), return `true`, otherwise false. If true, the scheduler will then assume that the picked task is still running, which cuts off "useless" paths where a completed task is polled again, and is even necessary to make things terminate, if nondeterminism is involved: if we pick the task nondeterministically, and don't have the restriction to still running tasks, we could poll the same task over and over again. However, for most deterministic scheduling strategies, e.g. the round robin scheduling strategy, assuming that the picked task is still running is generally not possible because if that task has ended, we are saying `assume(false)` and the verification effectively stops (which is undesirable, of course).

celinval replied:
I think you should document that. :)

fzaiser replied:
Updated the docs.

celinval commented on library/kani/src/futures.rs:
```diff
@@ -41,3 +44,216 @@ const NOOP_RAW_WAKER: RawWaker = {
 };
+
+static mut EXECUTOR: Scheduler = Scheduler::new();
+const MAX_TASKS: usize = 16;
```
Can we make this a parameter of the Scheduler?

fzaiser replied:
I guess we could, but then it would have to be an argument of `spawnable_block_on` as well. Given that anything with more nondeterminism than 16 tasks is unlikely to terminate in a reasonable time frame anyway, I don't think I see the point. Could you elaborate?

celinval replied:
Sure. This can actually affect the result of the verification, and none of our APIs are reflecting that.

Quick questions to help me understand this:
1- What will happen if by any chance we hit the MAX?
2- Does using a lower MAX help with the verification time?

celinval replied:
If lowering may improve the verification time, wouldn't it be helpful for users to be able to set this value?

fzaiser replied:
Well, as soon as the `Vec` performance improvements land, I'd like to replace this by a vector anyway. This is more flexible and does not require a constant upper bound. Therefore, I'd rather not expose the maximum number of tasks in the API.

fzaiser replied:
Now using `Vec` instead of an array, so this is resolved.

celinval commented on library/kani/src/futures.rs:
```diff
@@ -41,3 +44,216 @@ const NOOP_RAW_WAKER: RawWaker = {
+    /// Runs the scheduler with the given scheduling plan until all tasks have completed
+    #[inline] // to work around linking issue
+    fn run(&mut self, mut scheduling_plan: impl SchedulingStrategy) {
+        let waker = unsafe { Waker::from_raw(NOOP_RAW_WAKER) };
```
what does this mean? No waker?

fzaiser replied:
It is a waker, but it does nothing. That's OK because the scheduler does not rely on wakers. If we implemented wakers as in tokio, that would mean having another queue that slows things down.

celinval replied:
Can we use https://docs.rs/futures/latest/futures/task/fn.noop_waker.html instead?

fzaiser replied:
We could, but it would mean pulling in a dependency just for this one type, which isn't very hard to define ourselves. Also, by defining it ourselves, we retain the option of optimizing it for Kani in the future.

celinval replied:
fair. Thanks

celinval reviewed (commented):
More tests please! :)

fzaiser commented:
TODO: add check that the number of tasks is less than `MAX_TASKS`.

fzaiser commented:
With the MIR linker, this is no longer blocked! Ready for review @danielsn @celinval

fzaiser commented:
The latest CI run failed for `spawn.rs`: `CBMC failed with status 137`. Exit code 137 means that it ran out of memory, I think.

@celinval: could this be related to the new linker? Before this PR, all the scheduling code was in the file to be verified and it finished quickly. Now that it is in the library, it takes a long time.

celinval commented:
Yes,

> The latest CI run failed for `spawn.rs`: `CBMC failed with status 137`. Exit code 137 means that it ran out of memory, I think.
> 
> @celinval: could this be related to the new linker? If the scheduling code is in the same file, like in `manual_spawn.rs`, everything finishes quickly. (#1818 seems to be related?)

Yes, the main purpose of the new linker was to get rid of the undefined symbols from the standard library. Fixing this issue meant that some models got bigger and harder to process.

fzaiser commented:
> Yes, the main purpose of the new linker was to get rid of the undefined symbols from the standard library. Fixing this issue meant that some models got bigger and harder to process.

@celinval That makes sense. What I don't get is that it works fine if all the scheduling code is in the same file, see [`manual_spawn.rs`](https://github.com/model-checking/kani/blob/main/tests/kani/AsyncAwait/manual_spawn.rs). This file uses the same code that this PR puts into the Kani library, but puts that code directly into the file with the harness. That seems to verify much more quickly, which does not make sense to me. Do you know why?

celinval commented:
> > Yes, the main purpose of the new linker was to get rid of the undefined symbols from the standard library. Fixing this issue meant that some models got bigger and harder to process.
> 
> @celinval That makes sense. What I don't get is that it works fine if all the scheduling code is in the same file, see [`manual_spawn.rs`](https://github.com/model-checking/kani/blob/main/tests/kani/AsyncAwait/manual_spawn.rs). This file uses the same code that this PR puts into the Kani library, but puts that code directly into the file with the harness. That seems to verify much more quickly, which does not make sense to me. Do you know why?

I see. Let me take a look, but it is possible that moving the file into a different crate changed how rustc optimizes the code.

celinval commented:
I looked at the issue you mentioned, and I think the difference is due to the second harness in `spawn.rs` (`nondeterministic_schedule`). If you remove it, the time and problem size is basically the same independent on the location of the scheduling code.

The interesting thing is that the time is pretty bad even if you run `--harness deterministic_schedule`. That's probably because we perform reachability analysis for all harnesses at the same time, instead of doing them separately. My guess is that CBMC slicer is not smart enough to prune the closures out of the model due to a more conservative pointer analysis.

You can see a similar effect with the `--legacy-linker` by just moving the scheduling code to `spawn.rs`. If you run `kani --harness deterministic_schedule --legacy-linker` with and without the second harness co-located in `spawn.rs`, the times are drastically different. In my machine, Kani verification succeeds after 70ish seconds if you remove the second harness, but it doesn't finish under 5min if you add the second one.

fzaiser commented:
@celinval Thanks for that investigation. I've removed the nondeterministic test for now to make CI (hopefully) pass. I've also converted a `bool` to an enum and improved the documentation.

fzaiser commented:
Hm, seems like CI still runs out of memory. I'm not sure what to do about that. Any ideas?

celinval commented:
Can you run it locally and measure the memory consumption so we have an idea on how bad the problem is?

celinval commented on tests/kani/AsyncAwait/spawn.rs:
```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//
+// compile-flags: --edition 2018
+// kani-flags: --enable-unstable --mir-linker
```
btw, you don't need these flags any more. The MIR Linker is now the default one.

fzaiser replied:
A lot of tests still seem to have it. But I'll remove it.

celinval replied:
We haven't had time to remove them yet, but I think we should avoid adding them. We will probably fix the old tests once we remove the `--mir-linker` option which today is a no-op.

fzaiser replied:
A lot of tests still seem to have it. But I'll remove it.

fzaiser commented:
> Can you run it locally and measure the memory consumption so we have an idea on how bad the problem is?

@celinval, I just tried that with the second harness (the nondeterministic one) completely removed, so #1855 does not apply. It OOMed after 1 hour, maximum resident set size was about 15 GB. (My computer has 16 GB RAM, so it took up almost all the available memory.)

Should I open a bug about this? It makes no sense to me that it would take so much longer just because the code is in the Kani library instead of the same file. What do you think? cc @danielsn

celinval commented:
> > Can you run it locally and measure the memory consumption so we have an idea on how bad the problem is?
> 
> @celinval, I just tried that with the second harness (the nondeterministic one) completely removed, so #1855 does not apply. It OOMed after 1 hour, maximum resident set size was about 15 GB. (My computer has 16 GB RAM, so it took up almost all the available memory.)
> 
> Should I open a bug about this? It makes no sense to me that it would take so much longer just because the code is in the Kani library instead of the same file. What do you think? cc @danielsn

Sure. You can open a bug.

fzaiser commented:
Submitted #1876. This PR is **blocked** on this issue.

fzaiser commented:
The test is passing! Now that vectors are faster, I want to change this to use vectors instead of arrays before merging.

fzaiser commented:
@danielsn @celinval This should be ready for another round of review. Note that we now use vectors instead of arrays because they have gotten fast enough. Two questions to decide before merging:

1. The scheduling strategies `NondeterministicScheduling` and `NondetFairScheduling` are currently untested (commented out) because they take too long. (I stopped them after half an hour and taking 15 GB of RAM.) Should we make them private given that they're untested? I would prefer to leave them in (and maybe hide their documentation), so users can still experiment with them if they absolutely want to.
2. Should we remove `manual_spawn.rs`? Since the two versions now take almost no time (but see #1876), it may be redundant to have both.

celinval commented:
Our libraries don't usually have much logic into them, so there's not much to test. In these news APIs, I think we should test the scheduler logic. We should either have proofs for them or just concrete tests. An error in the scheduler could result in a soundness issue, right?

fzaiser commented:
@celinval: There is a proof harness using the scheduler (with the `RoundRobin` strategy) in `spawn.rs`. What additional tests or properties to prove do you have in mind?

The two nondeterministic scheduling strategies can only be tested in Kani proof harnesses, not in unit tests (because they use `kani::any()`). But even simple proof harnesses using them take a long time and lots of RAM. Altogether, this means they should not be included if you want them to be tested. Did I understand this correctly? (Another option to consider would be to include them, but hide their documentation. So users could still experiment with them, but we'd consider them "unstable".)

celinval commented:
Can you take advantage of the concrete playback infrastructure for injecting concrete values for testing purpose?

I do worry about the scalability. If we can't write simple harnesses with them, what value do we get from adding them to the Kani library at this point?

fzaiser commented:
That's a good point. I have removed the nondeterministic scheduling strategies. No only the `RoundRobin` strategy is included. (And there's a test for it.)

tautschnig commented:
> That's a good point. I have removed the nondeterministic scheduling strategies. No only the `RoundRobin` strategy is included. (And there's a test for it.)

But wouldn't the non-deterministic schedules be required for soundness? They might be impractical right now, but perhaps there is a way to make them tractable in future?

celinval commented:
> > That's a good point. I have removed the nondeterministic scheduling strategies. No only the `RoundRobin` strategy is included. (And there's a test for it.)
> 
> But wouldn't the non-deterministic schedules be required for soundness? They might be impractical right now, but perhaps there is a way to make them tractable in future?

Users can still create non-deterministic schedules and I think we should still keep an implementation in our benchmarks to know when they become practical, so we know it's time to add them.

I just don't think they should be in the Kani library until then. Specially since it cannot be tested at all today.

celinval commented on tests/ui/missing-function/replaced-description/expected:
```diff
@@ -1,3 +1,3 @@
 Status: FAILURE\
-Description: "Function `<alloc::string::String as core::clone::Clone>::clone` with missing definition is unreachable"
```
Why did this change?

fzaiser replied:
I believe because the Kani library now uses a vector (for storing the task list) and it just so happens to be checked before the String's clone method.

celinval commented on tests/kani/AsyncAwait/spawn.rs:
```diff
@@ -0,0 +1,30 @@
+
+#[kani::proof]
+#[kani::unwind(4)]
+fn deterministic_schedule() {
```
This is not necessarily testing that the schedule is deterministic though.

fzaiser replied:
Now, but it tests the behavior under a deterministic schedule. I'm not sure what you think should be improved.

celinval replied:
I want to make sure what we are shipping in the Kani crate follows its spec. There is no test in this PR that actually ensures that the RoundRobin scheduler is in fact RoundRobin.

These tests are internal to the tool, and adding extra checks here doesn't mean we have to always keep the same order. But we ensure that our implementation is correct.

Also, if we do pick one specific order, that should be clear to the user.

fzaiser replied:
Alright, I'll add asserts to check this and change the name of the harness

celinval commented on tests/kani/AsyncAwait/spawn.rs:
```diff
@@ -0,0 +1,30 @@
+        async move {
+            let x3 = x2.clone();
+            kani::spawn(async move {
+                x3.fetch_add(1, Ordering::Relaxed);
```
If I understood this correctly, this should always runs first. In which case, you could assert that x3 is 0, right?

fzaiser replied:
Well, it depends how exactly the round robin is implemented, i.e. in what order tasks are added to the task list. I'm not sure we want to make any guarantees other than that it's deterministic

fzaiser replied:
As per discussion above, added the assert.

celinval commented on tests/kani/AsyncAwait/spawn.rs:
```diff
@@ -0,0 +1,30 @@
+                x3.fetch_add(1, Ordering::Relaxed);
+            });
+            kani::yield_now().await;
+            x2.fetch_add(1, Ordering::Relaxed);
```
and here you can assert that x2 is 1.

fzaiser replied:
Done

celinval commented on tests/kani/AsyncAwait/manual_spawn.rs:
```diff
@@ -60,60 +110,98 @@ pub struct RoundRobin {
 }
 
 impl SchedulingStrategy for RoundRobin {
```
can you use the `kani::RoundRobin` instead?

fzaiser replied:
I thought it would be nice to have a manual implementation (not from the library) in `manual_spawn.rs` because there is a speed difference (#1876). But I can remove the file (since it duplicates code from the Kani library) if you want.

celinval replied:
You can add manual tests, but I would avoid a complete duplicate.

fzaiser replied:
I'm not sure what else to test manually, so I'll remove it.

celinval commented on tests/kani/AsyncAwait/manual_spawn.rs:
```diff
@@ -2,22 +2,45 @@
+///
+/// Note that spawn is not supported with this function. Use [`spawnable_block_on`] if you need it.
+// TODO: Give an error if spawn is used in the future passed to this function.
+pub fn block_on<T>(mut fut: impl Future<Output = T>) -> T {
```
why not `kani::block_on`?

fzaiser replied:
see https://github.com/model-checking/kani/pull/1659#discussion_r1104463102

fzaiser replied:
removed as discussed above

celinval commented on tests/kani/AsyncAwait/manual_spawn.rs:
```diff
@@ -32,25 +55,52 @@ const NOOP_RAW_WAKER: RawWaker = {
+///
+/// Finally, you have the option of providing your own scheduling strategy by implementing this trait.
+/// This can be useful, for example, if you want to verify that things work correctly for a very specific task ordering.
 pub trait SchedulingStrategy {
```
`kani::SchedulingStrategy`?

fzaiser replied:
see https://github.com/model-checking/kani/pull/1659#discussion_r1104463102

fzaiser replied:
removed as discussed above

celinval commented on tests/kani/AsyncAwait/manual_spawn.rs:
```diff
@@ -201,3 +290,22 @@ fn arc_spawn_deterministic_test() {
+
+// #[kani::proof]
+// #[kani::unwind(4)]
+// fn nondeterministic_schedule() {
```
Can you uncomment this and select the harness via kani-flags?

fzaiser replied:
done

celinval commented on tests/kani/AsyncAwait/manual_spawn.rs:
```diff
@@ -160,32 +247,34 @@ pub fn spawnable_block_on<F: Future<Output = ()> + Sync + 'static>(
+fn deterministic_schedule() {
     let x = Arc::new(AtomicI64::new(0)); // Surprisingly, Arc verified faster than Rc
     let x2 = x.clone();
     spawnable_block_on(
```
isn't this duplicating the harness from `spawn.rs`?

fzaiser replied:
see https://github.com/model-checking/kani/pull/1659#discussion_r1104463102

fzaiser replied:
removed as discussed above

celinval commented on tests/kani/AsyncAwait/manual_spawn.rs:
```diff
@@ -201,3 +290,22 @@ fn arc_spawn_deterministic_test() {
+
+// #[kani::proof]
+// #[kani::unwind(4)]
+// fn nondeterministic_schedule() {
```
Can you uncomment this and don't select the harness for now?

Note for self: This would be a great candidate for a regression that tracks intractable problems.

fzaiser replied:
I did this as well.

celinval commented on tests/kani/AsyncAwait/manual_spawn.rs:
```diff
@@ -201,3 +290,22 @@ fn arc_spawn_deterministic_test() {
+//         async move {
+//             let x3 = x2.clone();
+//             kani::spawn(async move {
+//                 x3.fetch_add(1, Ordering::Relaxed);
```
Maybe augment this to cover the scenario where this is run first and run second.

fzaiser replied:
I'm not sure what you mean here.

fzaiser replied:
File is deleted now anyway, so resolving this.

celinval commented on tests/kani/AsyncAwait/manual_spawn.rs:
```diff
@@ -201,3 +290,22 @@ fn arc_spawn_deterministic_test() {
+//                 x3.fetch_add(1, Ordering::Relaxed);
+//             });
+//             kani::yield_now().await;
+//             x2.fetch_add(1, Ordering::Relaxed);
```
ditto

fzaiser replied:
Not sure what you mean here.

fzaiser replied:
File is deleted now anyway, so resolving this.

rahulku commented:
Hi folks, what is the status of this PR?

fzaiser commented:
@rahulku I'll address the comments in the next few days.

fzaiser commented:
@celinval Thanks for your patience. Please take another look when you have time.

fzaiser commented:
@celinval I renamed the function `spawnable_block_on` to `block_on_with_spawn` as I think this name is clearer. ("spawnable" sounds like "can be spawned" instead of "can spawn")

fzaiser commented:
@dsn @celinval friendly reminder :)

celinval commented:
> @dsn @celinval friendly reminder :)

Sorry about that. Let me take a look

celinval commented:
Hi @fzaiser, I think this is good to go, except that I think we should add a mechanism to guard this code under a `unstable` flag. Let me just sketch something quick and we can merge this. Thanks!

celinval commented:
@fzaiser I created https://github.com/model-checking/kani/pull/2281 to get some feedback. Feel free to add comments there, but it should be a very quick change (famous last words).

celinval commented:
I created https://github.com/model-checking/kani/pull/2373 to add the unstable attribute and I'm planning to create another one to enforce the opt-in model.

Once https://github.com/model-checking/kani/pull/2373 is merged, I believe you can just tag the public APIs from this PR as unstable and we can merge these changes.

celinval commented:
Hi @fzaiser, I just wanted to let you know that we have merged the unstable API code, so you should be good to go.

fzaiser commented:
@celinval This should be good to go :) Could you please merge this if you're happy with it?

zhassan-aws commented:
> I'm not sure what the perf-benchcomp failure is about. Could it be spurious? Any ideas?

Yes, this is most likely spurious due to variance in performance. The `perf-benchcomp` workflow is not blocking for merges.

celinval reviewed (approved):
Thanks! Can you please add a reason to the unstable attribute? It could be something "recently added" or something like "experimental API". @adpaco-aws any suggestion?

adpaco-aws commented:
> Thanks! Can you please add a reason to the unstable attribute? It could be something "recently added" or something like "experimental API". @adpaco-aws any suggestion?

I'd suggest "experimental async support/library" or something along those lines.

Also, thanks @fzaiser for persistence on this one!

fzaiser commented:
@celinval Thanks for reviewing and making the final changes for the merge! 🎉
